torch_xla/distributed/xla_backend.py

```diff
@@ -0,0 +1,290 @@
+    self.timeout = timeout
+    self._mesh = []
+
+  def _get_reduce_type(self, reduce_op):
+    if reduce_op == dist.ReduceOp.SUM:
+      return xm.REDUCE_SUM
+    elif reduce_op == dist.ReduceOp.PRODUCT:
+      return xm.REDUCE_MUL
+    elif reduce_op == dist.ReduceOp.BAND:
+      return xm.REDUCE_AND
+    elif reduce_op == dist.ReduceOp.BOR:
+      return xm.REDUCE_OR
+    elif reduce_op == dist.ReduceOp.MIN:
+      return xm.REDUCE_MIN
+    elif reduce_op == dist.ReduceOp.MAX:
+      return xm.REDUCE_MAX
+    elif reduce_op == dist.ReduceOp.BXOR:
+      raise NotImplementedError(f'reduce op {reduce_op}')
+    else:
+      raise ValueError(f'Invalid reduce op {reduce_op}')
+
+  def allreduce(self, tensors, all_reduce_options):
+    reduce_type = self._get_reduce_type(all_reduce_options.reduceOp)
+
+    # TODO(hjm-aws): implement all_reduce_options.timeout.
+    xm.all_reduce(reduce_type, tensors, groups=self._mesh)
+    return WorkXla(tensors)
+
+  def allgather(self, output_tensors_list, input_tensors):
+    for input_tensor, output_tensors in zip(input_tensors, output_tensors_list):
+      result = xm.all_gather(input_tensor, groups=self._mesh)
+      for i, slice in enumerate(torch.split(result, input_tensor.shape[0])):
+        output_tensors[i].copy_(slice)
+
+    return WorkXla([t for sublist in output_tensors_list for t in sublist])
+
+  # Call site:
+  # https://github.com/pytorch/pytorch/blob/release/1.10/torch/distributed/distributed_c10d.py#L1129
+  def broadcast(self, tensors, opts):
+    root_tensor = tensors[opts.rootTensor]
+    root_rank = opts.rootRank
+    if root_rank != self.rank():
+      with torch.no_grad():
+        root_tensor.zero_()
+    xm.all_reduce(xm.REDUCE_SUM, [root_tensor], groups=self._mesh)
+
+    return WorkXla([root_tensor])
+
+  # Call site:
+  # https://github.com/pytorch/pytorch/blob/release/1.10/torch/distributed/distributed_c10d.py#L2355
+  def reduce_scatter(self, output_tensors, input_tensors_list, opts):
+    for input_tensors, output_tensor in zip(input_tensors_list, output_tensors):
+      # Ensure all inputs have the same shape.
+      first_shape = input_tensors[0].shape
+      for i, t in enumerate(input_tensors[1:]):
+        if first_shape != t.shape:
+          raise ValueError(f"Input {i+1}'s shape is different from input 0: "
+                           f"{t.shape} vs {first_shape}")
+      input_tensor = torch.cat(input_tensors)
+      reduce_type = self._get_reduce_type(opts.reduceOp)
+      groups = self._mesh
+      shard_count = len(groups[0]) if groups else self.size()
+      xm.reduce_scatter(
+          reduce_type,
+          input_tensor,
+          scatter_dim=0,
+          shard_count=shard_count,
+          scale=1,
+          groups=groups,
+          output=output_tensor)
+
+    return WorkXla(output_tensors)
+
+  # Call site:
+  # https://github.com/pytorch/pytorch/blob/70f57bcb1e45d21532bdb1c44d3aab018d1cbe88/torch/distributed/distributed_c10d.py#L2683
```

Comment:
Replace the reference link with `https://github.com/pytorch/pytorch/blob/release/1.11/torch/distributed/distributed_c10d.py#L2774`

Reply:
I think it is OK, this pr was already submitted to master and we don't need to replace links for every release branch.